### PR TITLE
prov/tcp adding fi_reject support

### PR DESCRIPTION
**Pull request by vkrishna**
Signed-off-by: Venkata Krishna Nimmagadda <venkata.krishna.nimmagadda@intel.com>

### CODE REVIEW

File: prov/tcp/src/tcpx_conn_mgr.c [hunk 1]
```diff
@@ -128,14 +129,23 @@ static int poll_fds_add_item(struct poll_fd_mgr *poll_mgr,
...
+		poll_mgr->poll_fds[poll_mgr->nfds].events = POLLOUT;
+		poll_mgr->poll_fds[poll_mgr->nfds].revents = 0;
 		break;
 	case PASSIVE_SOCK:
 		tcpx_pep = container_of(poll_info->fid, struct tcpx_pep,
 					util_pep.pep_fid.fid);
 
 		poll_mgr->poll_fds[poll_mgr->nfds].fd = tcpx_pep->sock;
 		poll_mgr->poll_fds[poll_mgr->nfds].events = POLLIN;
+		poll_mgr->poll_fds[poll_mgr->nfds].revents = 0;
```
Comment (shefty) [inline prov/tcp/src/tcpx_conn_mgr.c]: revents = 0 is set in all possible cases (I think), just move outside switch statement

File: prov/tcp/src/tcpx_ep.c [hunk 2]
```diff
@@ -651,6 +651,54 @@ static int tcpx_pep_listen(struct fid_pep *pep)
...
+	struct poll_fd_info *fd_info;
+	struct tcpx_conn_handle *tcpx_handle;
+	struct tcpx_fabric *tcpx_fabric;
+	struct tcpx_pep *tcpx_pep = container_of(pep, struct tcpx_pep,
+						 util_pep.pep_fid);
+
+	tcpx_fabric = container_of(tcpx_pep->util_pep.fabric,
+				   struct tcpx_fabric, util_fabric);
+
+	tcpx_handle = calloc(1, sizeof(*tcpx_handle));
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: We should avoid memory allocations in the reject case.  Move any necessary memory allocations to the reporting of the connection.

File: prov/tcp/src/tcpx_conn_mgr.c [hunk 3]
```diff
@@ -128,14 +129,23 @@ static int poll_fds_add_item(struct poll_fd_mgr *poll_mgr,
...
-
+		poll_mgr->poll_fds[poll_mgr->nfds].revents = 0;
+		break;
+	case REJECT_SOCK:
+		handle = container_of(poll_info->fid,
+					   struct tcpx_conn_handle,
+					   handle);
+		poll_mgr->poll_fds[poll_mgr->nfds].fd = handle->conn_fd;
+		poll_mgr->poll_fds[poll_mgr->nfds].events = POLLOUT;
+		poll_mgr->poll_fds[poll_mgr->nfds].revents = 0;
```
Comment (shefty) [inline prov/tcp/src/tcpx_conn_mgr.c]: I'm not following the reject handling.  If the app rejects a connection, we should close the socket and clean up.  Why is it being handled by the polling manager?

**Comment by vkrishna** [comment]
Makes sense.  It seems like I cooked up something complex for seemingly straight forward case :)

File: prov/tcp/src/tcpx_ep.c [hunk 4]
```diff
@@ -648,7 +648,30 @@ static int tcpx_pep_listen(struct fid_pep *pep)
...
+	hdr.version = OFI_CTRL_VERSION;
+	hdr.type = ofi_ctrl_nack;
+	hdr.seg_size = paramlen;
 
+	ofi_send_socket(tcpx_handle->conn_fd, &hdr,
+			sizeof(hdr), MSG_NOSIGNAL);
+
+	if (paramlen)
+		ofi_send_socket(tcpx_handle->conn_fd, param,
+				paramlen, MSG_NOSIGNAL);
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: I don't think NOSIGNAL is needed here or above.  I'm not sure if this is portable.

**Comment by vkrishna** [comment]
updated the commit with that change.

**Pull request merged**